dynamic_instrumentation/src/instrument.rs

```diff
@@ -123,6 +129,198 @@ fn is_region_or_unsafe_ptr(ty: Ty) -> bool {
+        // block
+        for (bid, block) in body.basic_blocks().iter_enumerated().rev() {
+            for (sid, statement) in block.statements.iter().enumerate().rev() {
+                if let StatementKind::Assign(stmt) = &statement.kind {
+                    let (ref place, _) = **stmt;
+                    if place
+                        .as_local()
+                        .filter(|l| self.address_taken.contains(l))
+                        .is_some()
+                    {
+                        // put just below first assignment
+                        local_address_statements.push((*place, bid, sid + 1));
+                    }
+                }
+            }
+            if let Some(term) = &block.terminator {
+                match &term.kind {
+                    TerminatorKind::Call {
+                        func: _func,
+                        args: _args,
+                        destination,
+                        target,
+                        ..
+                    } => {
+                        if destination
+                            .as_local()
+                            .filter(|l| self.address_taken.contains(l))
+                            .is_some()
+                        {
+                            if let Some(next_block) = target {
+                                // put into first statement of following block
+                                local_address_statements.push((*destination, *next_block, 0));
+                            }
+                        }
+                    }
+                    TerminatorKind::DropAndReplace {
+                        place,
+                        value,
+                        target,
+                        unwind: _,
+                    } if value.place().is_some() => {
+                        if let Some(local_substitute) =
+                            place.as_local().and_then(|l| self.local_substitute.get(&l))
+                        {
+                            // put into first statement of following block
+                            local_address_statements.push((*place, *target, 0));
+                            self.local_substitute.insert(place.local, *local_substitute);
+                        }
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        // END ADDRESS-TAKING STATEMENTS
+        // visit places first before inserting `_y = &raw _x`, so that we don't accidentally
+        // rewrite `_y = &raw _x` into `y = &raw (*_y)`
+        self.super_body(body);
+
+        // sort the insertion locations by statement id descending so that statement insertion
+        // for statements with indices N and M, where N < M, does not shift/invalidate index M
+        local_address_statements.sort_by_key(|(_, _bid, sid)| *sid);
+        let local_address_statements = local_address_statements.into_iter().rev();
```

Comment:
It would be cleaner and faster to sort by `-*sid` here, or (if this tuple becomes a struct) define `PartialOrd` accordingly.

Reply:
What do you mean exactly?  By the faster part?  The sort needs to be done in place, right?

Reply:
Ah, yeah, never mind the faster bit--I forgot we were just reversing an iterator, because `local_address_statements` changes from a concrete collection into an iterator with this rebinding.

Reply:
Yeah, the `.rev()` should be zero-cost, and if `sid` is unsigned, `-sid` is a problem.  There's `Ordering::reverse`, but that would only work with `.sort_by()`, not `.sort_by_key()`.